tests/magic_spell_test.cpp

```diff
@@ -0,0 +1,105 @@
+{
+    spell test_spell( sp_id );
+    test_spell.set_level( from_level );
+    return test_spell.exp_to_next_level();
+}
+
+TEST_CASE( "experience to level up spell", "[magic][spell][xp]" )
+{
+    spell_id lava_id( "test_spell_lava" );
+
+    SECTION( "experience needed to level up spells" ) {
+        CHECK( spell_xp_to_level( lava_id, 0 ) == 4881 );
+        CHECK( spell_xp_to_level( lava_id, 1 ) == 1751 );
+        CHECK( spell_xp_to_level( lava_id, 2 ) == 2027 );
+        CHECK( spell_xp_to_level( lava_id, 3 ) == 2347 );
+        CHECK( spell_xp_to_level( lava_id, 4 ) == 2717 );
+        CHECK( spell_xp_to_level( lava_id, 5 ) == 3147 );
+        CHECK( spell_xp_to_level( lava_id, 6 ) == 3644 );
+        CHECK( spell_xp_to_level( lava_id, 7 ) == 4220 );
+        CHECK( spell_xp_to_level( lava_id, 8 ) == 4886 );
+        CHECK( spell_xp_to_level( lava_id, 9 ) == 5659 );
+        CHECK( spell_xp_to_level( lava_id, 10 ) == 6552 );
+        CHECK( spell_xp_to_level( lava_id, 11 ) == 7587 );
+        CHECK( spell_xp_to_level( lava_id, 12 ) == 8785 );
+        CHECK( spell_xp_to_level( lava_id, 13 ) == 10173 );
+        CHECK( spell_xp_to_level( lava_id, 14 ) == 11780 );
+        CHECK( spell_xp_to_level( lava_id, 15 ) == 13641 );
+        CHECK( spell_xp_to_level( lava_id, 16 ) == 15795 );
+        CHECK( spell_xp_to_level( lava_id, 17 ) == 18291 );
+        CHECK( spell_xp_to_level( lava_id, 18 ) == 21180 );
+        CHECK( spell_xp_to_level( lava_id, 19 ) == 24525 );
+    }
+}
+
+// Return spell damage at a given level
+static int spell_damage( const spell_id test_spell_id, const int spell_level )
+{
+    spell test_spell( test_spell_id );
+    test_spell.set_level( spell_level );
+    return test_spell.damage();
+}
+
+TEST_CASE( "spell damage", "[magic][spell][damage]" )
+{
+    spell_id sp_id( "test_spell_pew" );
+    spell_type sp_type = sp_id.obj();
+
+    // Damage varies from 1 (at level 0) to a maximum possible of 5
+    REQUIRE( sp_type.min_damage == 1 );
+    REQUIRE( sp_type.max_damage == 5 );
+
+    // Damage goes up this much each level (using whole number for more consistent testing)
+    REQUIRE( sp_type.damage_increment == 1.0f );
+
+    // Highest level spell can reach (though maximum damage is reached before then)
+    REQUIRE( sp_type.max_level == 10 );
+
+    SECTION( "spell damage is min_damage + (level x damage_increment), up to max_damage" ) {
+        CHECK( spell_damage( sp_id, 0 ) == 1 );
+        CHECK( spell_damage( sp_id, 1 ) == 2 );
+        CHECK( spell_damage( sp_id, 2 ) == 3 );
+        CHECK( spell_damage( sp_id, 3 ) == 4 );
+        CHECK( spell_damage( sp_id, 4 ) == 5 );
+        CHECK( spell_damage( sp_id, 5 ) == 5 );
+        CHECK( spell_damage( sp_id, 9 ) == 5 );
+        // FIXME: Tests hang when going to maximum level
+        //CHECK( spell_damage( sp_id, 10 ) == 5 );
+    }
+}
+
+TEST_CASE( "spell area of effect", "[magic][spell][aoe]" )
```

Comment:
if you're gonna exercise how aoe gets tripoints, you might need 3 separate tests for cone, line, and "circle"

Reply:
Thanks, I'll keep it in mind. I had only looked at the spell class functions `src/magic.cpp` so far, but I see the effect functions defined in `src/magic_spell_effect.cpp` will definitely come into play here too.